test/async-action.ts

```diff
@@ -214,6 +214,192 @@ test("it should support logging", t => {
+
+    setTimeout(() => {
+        // TODO: mweh on any cast...
+        ;(x.f(/*test binding*/ 2) as any).then((v: number) => {
+            // note: ideally, type of v should be inferred..
+            t.is(v, 5)
+            t.deepEqual(values, [1, 2, 5])
+            t.is(x.a, 5) // correct instance modified?
+            t.end()
+        })
+    }, 10)
+})
+
+test("it should support asyncActionWithCancel(cb, type) as decorator for no-value variable", t => {
+    const values: any[] = []
+
+    mobx.useStrict(true)
+
+    class X {
+        @mobx.observable a = 1;
+
+        @utils.asyncActionWithCancel(() => {}, 'takeLatest')
+        noValueParam
+    }
+
+    const x = new X()
+    x.noValueParam = function *(initial: number) {
+        this.a = initial // this runs in action
+        try {
+            this.a = yield delay(100, 5, true) // and this as well!
+            yield delay(100, 0)
+            this.a = 4
+        } catch (e) {
+            this.a = e
+        }
+        return this.a
+    }
+
+    mobx.reaction(() => x.a, v => values.push(v), { fireImmediately: true })
+
+    setTimeout(() => {
+        // TODO: mweh on any cast...
+        ;(x.noValueParam(/*test binding*/ 2) as any).then((v: number) => {
+            // note: ideally, type of v should be inferred..
+            t.is(v, 5)
+            t.deepEqual(values, [1, 2, 5])
+            t.is(x.a, 5) // correct instance modified?
+            t.end()
+        })
+    }, 10)
+})
+
+test("it should support canceling the effects when call canceler", t => {
+    mobx.useStrict(true)
+    const values: any[] = []
+    const x = mobx.observable({ a: 1 })
+    mobx.reaction(() => x.a, v => values.push(v), { fireImmediately: true })
+
+    let canceler
+    const f = utils.asyncActionWithCancel(cancel => canceler = cancel)(function*(initial: number) {
+        x.a = initial // this runs in action
+        try {
+            x.a = yield delay(1000, 5, true) // and this as well!
+        } catch (e) {
+            x.a = e
+        }
+        return x.a
+    })
+
+    setTimeout(() => {
+        f(2)
+        setTimeout(() => {
+            canceler()
+            t.is(x.a, 2)
+            t.deepEqual(values, [1, 2])
+            t.end()
+        }, 150)
+    }, 10)
+})
+
+test("it should support canceling prev the same kinds of async generator actions in 'takeLatest' type", t => {
+    mobx.useStrict(true)
+    const values: any[] = []
+    const x = mobx.observable({ a: 1 })
+    mobx.reaction(() => x.a, v => values.push(v), { fireImmediately: true })
+
+    const f = utils.asyncActionWithCancel(() => {}, 'takeLatest')(function*(initial: number) {
+        x.a = initial // this runs in action
+        x.a = yield delay(100, 3) // and this as well!
+        yield delay(100, 0)
+        x.a = 4
+        return x.a
+    })
+
+    setTimeout(() => {
+        f(2)
+        f(3)
+        f(5).then((v: number) => {
```

Comment:
With `takeLatest`, do the original promises reject, or will they resolve with the same value as `f(5)`?

Reply:
Unfortunately not, since generator.prototype.return will break the program and return some value prepared by the developer. 
**It's the side-effect of the code process control by generator.**
```
f(2).then(v => console.log(v)) // canceled, depend on the value of gen.return, in my pr it's undefined
f(3).then(v => console.log(v)) // canceled, depend on the value of gen.return, in my pr it's undefined
f(5).then(v => console.log(v)) // as expected 4
```
For this reason, the idea of canceler is not quite adaptive.(lol)
Hope the rough idea can help you something.